File: ChatDemo-UI3.0/Class/Core Data/InvitationManager.m
/************************************************************
 *  * Hyphenate  
 * __________________
 * Copyright (C) 2016 Hyphenate Inc. All rights reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of Hyphenate Inc.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Hyphenate Inc.
 */

#import "InvitationManager.h"

@interface InvitationManager ()

@property (nonatomic, strong) NSUserDefaults *defaults;

@end

static InvitationManager *sharedInstance = nil;

@implementation InvitationManager


+ (instancetype)sharedInstance{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

-(instancetype)init{
    if (self = [super init]) {
        self.defaults = [NSUserDefaults standardUserDefaults];
    }
    
    return self;
}


#pragma mark - Data

- (void)addInvitation:(RequestEntity *)requestEntity loginUser:(NSString *)username
{
    NSData *defalutData = [self.defaults objectForKey:username];
    if (defalutData) {
        
        NSMutableArray *requests = [[NSKeyedUnarchiver unarchiveObjectWithData:defalutData] mutableCopy];
        if (requests) {
            [requests addObject:requestEntity];
            
            NSData *data = [NSKeyedArchiver archivedDataWithRootObject:requests];
            
            [self.defaults setObject:data forKey:username];
        }
    }
}

- (void)removeInvitation:(RequestEntity *)requestEntity loginUser:(NSString *)username
{
    NSData *defalutData = [self.defaults objectForKey:username];
    
    if (!defalutData) {
        return;
    }
    
    NSMutableArray *requests = [[NSKeyedUnarchiver unarchiveObjectWithData:defalutData] mutableCopy];
    
    RequestEntity *needDelete;
    for (RequestEntity *request in requests) {
        if ([request.groupId isEqualToString:requestEntity.groupId] &&
            [request.receiverUsername isEqualToString:requestEntity.receiverUsername]) {
            needDelete = request;
            break;
        }
    }
    [requests removeObject:needDelete];
    
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:requests];
    
    [self.defaults setObject:data forKey:username];
}

- (NSArray *)getSavedFriendRequests:(NSString *)username
{
    NSData *defalutData = [self.defaults objectForKey:username];
    
    NSArray *requestObjects = [[NSArray alloc] init];
    
    if (defalutData) {
        requestObjects = [NSKeyedUnarchiver unarchiveObjectWithData:defalutData];
    }
    
    return requestObjects;
}

@end


@interface RequestEntity ()<NSCoding>

@end

@implementation RequestEntity

-(void)encodeWithCoder:(NSCoder *)aCoder{
    [aCoder encodeObject:_applicantUsername forKey:@"applicantUsername"];
    [aCoder encodeObject:_applicantNick forKey:@"applicantNick"];
    [aCoder encodeObject:_reason forKey:@"reason"];
    [aCoder encodeObject:_receiverUsername forKey:@"receiverUsername"];
    [aCoder encodeObject:_receiverNick forKey:@"receiverNick"];
    [aCoder encodeObject:_style forKey:@"style"];
    [aCoder encodeObject:_groupId forKey:@"groupId"];
    [aCoder encodeObject:_groupSubject forKey:@"subject"];
}

-(id)initWithCoder:(NSCoder *)aDecoder{
    if(self = [super init]){
        _applicantUsername = [aDecoder decodeObjectForKey:@"applicantUsername"];
        _applicantNick = [aDecoder decodeObjectForKey:@"applicantNick"];
        _reason = [aDecoder decodeObjectForKey:@"reason"];
        _receiverUsername = [aDecoder decodeObjectForKey:@"receiverUsername"];
        _receiverNick = [aDecoder decodeObjectForKey:@"receiverNick"];
        _style = [aDecoder decodeObjectForKey:@"style"];
        _groupId = [aDecoder decodeObjectForKey:@"groupId"];
        _groupSubject = [aDecoder decodeObjectForKey:@"subject"];
        
    }
    
    return self;
}

@end

File: swift-demo/Hyphenate Messenger/Data Model/InvitationManager.swift
//
//  InvitationManager.swift
//  Hyphenate Messenger
//
//  Created by Peng Wan on 10/2/16.
//  Copyright © 2016 Hyphenate Inc. All rights reserved.
//

import Foundation
class InvitationManager:NSObject{
    
    var sharedInstance: InvitationManager? = nil
    var defaults: UserDefaults!
        
    static let sharedInstance = InvitationManager()
    
    override init() {
        super.init()
        self.defaults = UserDefaults.standard
        
    }
        // MARK: - Data
        
        func addInvitation(_ requestEntity: RequestEntity, loginUser username: String) {
            if let defalutData = self.defaults.object(forKey: username) as? Data {
                if var requests = NSKeyedUnarchiver.unarchiveObject(with: defalutData) as? [RequestEntity]{
                    requests.append(requestEntity)
                    let data = NSKeyedArchiver.archivedData(withRootObject: requests)
                    self.defaults.set(data, forKey: username)
                }
            } else {
                let data = NSKeyedArchiver.archivedData(withRootObject: [requestEntity])
                self.defaults.set(data, forKey: username)
            }
            DispatchQueue.main.async {
                NotificationCenter.default.post(name: Notification.Name(rawValue: "kNotification_requestUpdated"), object: nil)
            }
        }
    
        func removeInvitation(_ requestEntity: RequestEntity, loginUser username: String) {
            if let defalutData = self.defaults.object(forKey: username) as? Data{
                if var requests = NSKeyedUnarchiver.unarchiveObject(with: defalutData) as? [RequestEntity]{
                    var needDelete: RequestEntity?
                    for request: RequestEntity in requests {
                        if (request.groupId == requestEntity.groupId) && (request.applicantUsername == requestEntity.applicantUsername) {
                            needDelete = request
                            break
                        }
                    }
                    if let _ = needDelete{
                        if let index = requests.index(where: {$0 == needDelete!}){
                            requests.remove(at: index)
                            let data = NSKeyedArchiver.archivedData(withRootObject: requests)
                            self.defaults.set(data, forKey: username)
                            DispatchQueue.main.async {
                                NotificationCenter.default.post(name: Notification.Name(rawValue: "kNotification_requestUpdated"), object: nil)
                            }
                        }
                    }
                }
            }
        }
        
        func getSavedFriendRequests(_ username: String) -> [RequestEntity]? {
            if let defalutData = self.defaults.object(forKey: username) as? Data {
                return NSKeyedUnarchiver.unarchiveObject(with: defalutData) as? [RequestEntity]
            }
            return nil
        }
}
